File: CS/NodeChecking.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;

namespace DevExpress.XtraTreeList.Demos.Tutorials {
    /// <summary>
    /// Summary description for NodeChecking.
    /// </summary>
    public partial class NodeChecking : DevExpress.XtraEditors.XtraUserControl {
        public NodeChecking() {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();
            InitData();
            //
            // TODO: Add any constructor code after InitializeComponent call
            //
        }

        private void InitData() {
            DevExpress.XtraTreeList.Design.XViews xv = new DevExpress.XtraTreeList.Design.XViews(treeList1);
            treeList1.SelectImageList = null;
            try {
                SetCheckedNode(treeList1.Nodes[0].Nodes[0]);
            }
            catch { }
        }

        private CheckState GetCheckState(object obj) {
            if(obj != null) return (CheckState)obj;
            return CheckState.Unchecked;
        }
        //<treeList1>
        private void SetCheckedNode(DevExpress.XtraTreeList.Nodes.TreeListNode node) {
            CheckState check = GetCheckState(node.Tag);
            if(check == CheckState.Indeterminate || check == CheckState.Unchecked) check = CheckState.Checked;
            else check = CheckState.Unchecked;
            treeList1.BeginUpdate();
            node.Tag = check;
            SetCheckedChildNodes(node, check);
            SetCheckedParentNodes(node, check);
            treeList1.EndUpdate();
        }
        //</treeList1>

        private void SetCheckedChildNodes(DevExpress.XtraTreeList.Nodes.TreeListNode node, CheckState check) {
            for(int i = 0; i < node.Nodes.Count; i++) {
                node.Nodes[i].Tag = check;
                SetCheckedChildNodes(node.Nodes[i], check);
            }
        }
        private void SetCheckedParentNodes(DevExpress.XtraTreeList.Nodes.TreeListNode node, CheckState check) {
            if(node.ParentNode != null) {
                bool b = false;
                CheckState state;
                for(int i = 0; i < node.ParentNode.Nodes.Count; i++) {
                    if(node.ParentNode.Nodes[i].Tag == null) state = CheckState.Unchecked;
                    else state = (CheckState)node.ParentNode.Nodes[i].Tag;
                    if(!check.Equals(state)) {
                        b = !b;
                        break;
                    }
                }
                node.ParentNode.Tag = b ? CheckState.Indeterminate : check;
                SetCheckedParentNodes(node.ParentNode, check);
            }
        }

        //<treeList1>
        private void treeList1_GetStateImage(object sender, DevExpress.XtraTreeList.GetStateImageEventArgs e) {
            CheckState check = GetCheckState(e.Node.Tag);
            if(check == CheckState.Unchecked)
                e.NodeImageIndex = 0;
            else if(check == CheckState.Checked)
                e.NodeImageIndex = 1;
            else e.NodeImageIndex = 2;
        }

        private void treeList1_KeyDown(object sender, System.Windows.Forms.KeyEventArgs e) {
            if(e.KeyData == Keys.Space)
                SetCheckedNode(treeList1.FocusedNode);
        }

        private void treeList1_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e) {
            if(e.Button == MouseButtons.Left) {
                DevExpress.XtraTreeList.TreeListHitInfo hInfo = treeList1.CalcHitInfo(new Point(e.X, e.Y));
                if(hInfo.HitInfoType == HitInfoType.StateImage)
                    SetCheckedNode(hInfo.Node);
            }
        }
        //</treeList1>
    }
}

File: VB/NodeChecking.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data

Namespace DevExpress.XtraTreeList.Demos.Tutorials
	''' <summary>
	''' Summary description for NodeChecking.
	''' </summary>
	Partial Public Class NodeChecking
		Inherits DevExpress.XtraEditors.XtraUserControl
		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()
			InitData()
			'
			' TODO: Add any constructor code after InitializeComponent call
			'
		End Sub

		Private Sub InitData()
			Dim xv As New DevExpress.XtraTreeList.Design.XViews(treeList1)
			treeList1.SelectImageList = Nothing
			Try
				SetCheckedNode(treeList1.Nodes(0).Nodes(0))
			Catch
			End Try
		End Sub

		Private Function GetCheckState(ByVal obj As Object) As CheckState
			If obj IsNot Nothing Then
				Return CType(obj, CheckState)
			End If
			Return CheckState.Unchecked
		End Function
		'<treeList1>
		Private Sub SetCheckedNode(ByVal node As DevExpress.XtraTreeList.Nodes.TreeListNode)
			Dim check As CheckState = GetCheckState(node.Tag)
			If check = CheckState.Indeterminate OrElse check = CheckState.Unchecked Then
				check = CheckState.Checked
			Else
				check = CheckState.Unchecked
			End If
			treeList1.BeginUpdate()
			node.Tag = check
			SetCheckedChildNodes(node, check)
			SetCheckedParentNodes(node, check)
			treeList1.EndUpdate()
		End Sub
		'</treeList1>

		Private Sub SetCheckedChildNodes(ByVal node As DevExpress.XtraTreeList.Nodes.TreeListNode, ByVal check As CheckState)
			For i As Integer = 0 To node.Nodes.Count - 1
				node.Nodes(i).Tag = check
				SetCheckedChildNodes(node.Nodes(i), check)
			Next i
		End Sub
		Private Sub SetCheckedParentNodes(ByVal node As DevExpress.XtraTreeList.Nodes.TreeListNode, ByVal check As CheckState)
			If node.ParentNode IsNot Nothing Then
				Dim b As Boolean = False
				Dim state As CheckState
				For i As Integer = 0 To node.ParentNode.Nodes.Count - 1
					If node.ParentNode.Nodes(i).Tag Is Nothing Then
						state = CheckState.Unchecked
					Else
						state = CType(node.ParentNode.Nodes(i).Tag, CheckState)
					End If
					If (Not check.Equals(state)) Then
						b = Not b
						Exit For
					End If
				Next i
				If b Then
					node.ParentNode.Tag = CheckState.Indeterminate
				Else
					node.ParentNode.Tag = check
				End If
				SetCheckedParentNodes(node.ParentNode, check)
			End If
		End Sub

		'<treeList1>
		Private Sub treeList1_GetStateImage(ByVal sender As Object, ByVal e As DevExpress.XtraTreeList.GetStateImageEventArgs) Handles treeList1.GetStateImage
			Dim check As CheckState = GetCheckState(e.Node.Tag)
			If check = CheckState.Unchecked Then
				e.NodeImageIndex = 0
			ElseIf check = CheckState.Checked Then
				e.NodeImageIndex = 1
			Else
				e.NodeImageIndex = 2
			End If
		End Sub

		Private Sub treeList1_KeyDown(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles treeList1.KeyDown
			If e.KeyData = Keys.Space Then
				SetCheckedNode(treeList1.FocusedNode)
			End If
		End Sub

		Private Sub treeList1_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles treeList1.MouseDown
			If e.Button = MouseButtons.Left Then
				Dim hInfo As DevExpress.XtraTreeList.TreeListHitInfo = treeList1.CalcHitInfo(New Point(e.X, e.Y))
				If hInfo.HitInfoType = HitInfoType.StateImage Then
					SetCheckedNode(hInfo.Node)
				End If
			End If
		End Sub
		'</treeList1>
	End Class
End Namespace
